File: src/player/snapshot.rs
//! Cross-reload session snapshot codec.
//!
//! Port of `web/player-core/src/snapshot.ts` (BSD-3-Clause). Pure
//! serialize/validate/normalize; storage I/O and throttling stay host-side.
//!
//! v1 (shipped): `{v:1, items, index, positionSeconds, volume, normalizeMode}`.
//! v2 adds `repeat` and `shuffle`; `crossfadeSeconds` is accepted when it is
//! one of the shipped cycle values (4/8/12), otherwise 0.
//!
//! # Deliberate deviations from TypeScript
//!
//! - The Rust JSON parser rejects duplicate object keys; a payload with
//!   duplicates decodes as `None` (a safe no-restore), whereas
//!   `JSON.parse` would keep the last value. This is a robustness hardening,
//!   not a semantic change for well-formed snapshots.
//! - Numbers are emitted with Rust's shortest round-trip formatting rather
//!   than `JSON.stringify`'s; the output is shape-compatible JSON, not
//!   necessarily byte-identical.

use crate::json::{self, Value};

use super::types::{NormalizationMode, PlaybackItem, RepeatMode};

/// Current snapshot version.
pub const SNAPSHOT_VERSION: u32 = 2;
/// The version the web client originally shipped.
pub const SNAPSHOT_VERSION_V1: u32 = 1;

/// A normalized session snapshot.
#[derive(Debug, Clone, PartialEq)]
pub struct SessionSnapshot {
    /// Snapshot version (always [`SNAPSHOT_VERSION`] after decode).
    pub v: u32,
    /// Restorable items (raw JSON objects, preserved verbatim).
    pub items: Vec<Value>,
    /// Cursor index into `items`.
    pub index: i64,
    /// Position within the current track, in seconds.
    pub position_seconds: f64,
    /// Saved volume, when present.
    pub volume: Option<f64>,
    /// Saved normalization mode, when present.
    pub normalize_mode: Option<NormalizationMode>,
    /// Saved repeat policy (v1 payloads default to `Off`).
    pub repeat: RepeatMode,
    /// Saved shuffle flag (v1 payloads default to `false`).
    pub shuffle: bool,
    /// Crossfade seconds (0 = off; only 4/8/12 are honored).
    pub crossfade_seconds: f64,
}

/// Clamps an index into `0..length` (mirrors the historical restore clamp,
/// including `floor(NaN) || 0`).
pub fn clamp_index(index: f64, length: usize) -> i64 {
    let floored = index.floor();
    let value = if floored.is_nan() { 0.0 } else { floored };
    let hi = (length as f64 - 1.0).max(0.0);
    value.max(0.0).min(hi) as i64
}

fn number_at(v: &Value, key: &str) -> Option<f64> {
    match v.get(key) {
        Some(Value::Number(n)) => Some(*n),
        _ => None,
    }
}

fn is_restorable(item: &Value) -> bool {
    PlaybackItem::from_value(item).is_some()
}

/// Parses a v1 or v2 payload. Returns `None` when absent/corrupt/empty —
/// restoration then simply does not happen (never fatal).
pub fn decode_snapshot(raw: &str) -> Option<SessionSnapshot> {
    if raw.is_empty() {
        return None;
    }
    let value = json::parse(raw.as_bytes()).ok()?;
    // The version is validated but the decoded snapshot is normalized to the
    // current version (matching the TypeScript `v: SNAPSHOT_VERSION`).
    match number_at(&value, "v") {
        Some(v) if v == SNAPSHOT_VERSION as f64 || v == SNAPSHOT_VERSION_V1 as f64 => {}
        _ => return None,
    };
    let v = SNAPSHOT_VERSION;
    let items_value = value.get("items")?;
    let Value::Array(items) = items_value else {
        return None;
    };
    let items: Vec<Value> = items.iter().filter(|i| is_restorable(i)).cloned().collect();
    if items.is_empty() {
        return None;
    }
    let index = clamp_index(number_at(&value, "index").unwrap_or(0.0), items.len());
    let crossfade = number_at(&value, "crossfadeSeconds")
        .filter(|c| [4.0, 8.0, 12.0].contains(c))
        .unwrap_or(0.0);
    let normalize_mode = match value.get("normalizeMode") {
        Some(Value::String(s)) if matches!(s.as_str(), "off" | "album" | "track") => {
            Some(NormalizationMode::parse(s))
        }
        _ => None,
    };
    let repeat = match value.get("repeat") {
        Some(Value::String(s)) => RepeatMode::parse(s),
        _ => RepeatMode::Off,
    };
    let shuffle = matches!(value.get("shuffle"), Some(Value::Bool(true)));
    Some(SessionSnapshot {
        v,
        items,
        index,
        position_seconds: number_at(&value, "positionSeconds").unwrap_or(0.0),
        volume: number_at(&value, "volume"),
        normalize_mode,
        repeat,
        shuffle,
        crossfade_seconds: crossfade,
    })
}

/// Encodes a snapshot as JSON.
pub fn encode_snapshot(snapshot: &SessionSnapshot) -> String {
    let mut members: Vec<(String, Value)> = vec![
        ("v".into(), Value::Number(SNAPSHOT_VERSION as f64)),
        ("items".into(), Value::Array(snapshot.items.clone())),
        ("index".into(), Value::Number(snapshot.index as f64)),
        (
            "positionSeconds".into(),
            Value::Number(snapshot.position_seconds),
        ),
    ];
    if let Some(volume) = snapshot.volume {
        members.push(("volume".into(), Value::Number(volume)));
    }
    if let Some(mode) = snapshot.normalize_mode {
        members.push(("normalizeMode".into(), Value::String(mode.as_str().into())));
    }
    members.push((
        "repeat".into(),
        Value::String(snapshot.repeat.as_str().into()),
    ));
    members.push(("shuffle".into(), Value::Bool(snapshot.shuffle)));
    members.push((
        "crossfadeSeconds".into(),
        Value::Number(snapshot.crossfade_seconds),
    ));
    let mut out = String::new();
    write_value(&Value::Object(members), &mut out);
    out
}

/// Shortest round-trip JSON writer (finite → number, non-finite → `null`).
fn write_value(value: &Value, out: &mut String) {
    match value {
        Value::Null => out.push_str("null"),
        Value::Bool(b) => out.push_str(if *b { "true" } else { "false" }),
        Value::Number(n) => {
            if n.is_finite() {
                out.push_str(&format!("{n}"));
            } else {
                out.push_str("null");
            }
        }
        Value::String(s) => write_string(s, out),
        Value::Array(items) => {
            out.push('[');
            for (i, item) in items.iter().enumerate() {
                if i > 0 {
                    out.push(',');
                }
                write_value(item, out);
            }
            out.push(']');
        }
        Value::Object(members) => {
            out.push('{');
            for (i, (k, v)) in members.iter().enumerate() {
                if i > 0 {
                    out.push(',');
                }
                write_string(k, out);
                out.push(':');
                write_value(v, out);
            }
            out.push('}');
        }
    }
}

fn write_string(s: &str, out: &mut String) {
    out.push('"');
    for c in s.chars() {
        match c {
            '"' => out.push_str("\\\""),
            '\\' => out.push_str("\\\\"),
            '\n' => out.push_str("\\n"),
            '\r' => out.push_str("\\r"),
            '\t' => out.push_str("\\t"),
            '\u{08}' => out.push_str("\\b"),
            '\u{0c}' => out.push_str("\\f"),
            c if (c as u32) < 0x20 => out.push_str(&format!("\\u{:04x}", c as u32)),
            c => out.push(c),
        }
    }
    out.push('"');
}

File: web/player-core/src/snapshot.ts
// Cross-reload session snapshot codec (player-core M1; v2 in M6).
//
// Pure serialize/validate/normalize extracted from the web
// PlayerController.persist()/restore(). Storage I/O and throttling stay
// host-side; this module owns only the wire format and its invariants.
//
// Versions:
//   v1 (shipped): {v:1, items, index, positionSeconds, volume, normalizeMode}
//     — no playback policy; a restored session defaults to repeat 'off',
//       shuffle off.
//   v2 (M6): adds `repeat` ('off'|'one'|'all') and `shuffle` (boolean).
//
// The codec validates only what restoration relies on and clamps the rest;
// decode failures are non-fatal by design (best-effort resume).

import type { NormalizationMode } from './gain';
import type { RepeatMode } from './order';

export const SNAPSHOT_VERSION = 2;
/** The version the web client originally shipped. */
export const SNAPSHOT_VERSION_V1 = 1;

export interface SnapshotItemBase {
  /** Minimal identity contract for restoration; hosts may persist richer
   *  item shapes verbatim (the web persists QueueItems). */
  track?: { id?: number; audio?: { url?: string }; duration?: number };
  /** Modern item identity. Writers since the player-core extraction
   *  (c5bf447) always emit these; the restorer requires them so payloads
   *  persisted by PRE-extraction bundles (bare `{track,...}` items) are
   *  dropped wholesale instead of restoring unplayable queue entries that
   *  a later persist would write back over good storage. */
  id?: string;
  source?: { kind?: string; url?: string };
}

export interface SessionSnapshotV1 extends SnapshotItemBase {
  v: typeof SNAPSHOT_VERSION_V1;
  items: SnapshotItemBase[];
  index: number;
  positionSeconds: number;
  volume: number;
  normalizeMode: NormalizationMode;
}

export interface SessionSnapshot {
  v: typeof SNAPSHOT_VERSION;
  items: SnapshotItemBase[];
  index: number;
  positionSeconds: number;
  volume: number;
  normalizeMode: NormalizationMode;
  /** Playback policy (v2). Always present after decode (v1 → defaults). */
  repeat: RepeatMode;
  shuffle: boolean;
  /** Crossfade seconds; 0 = off. Optional for forward/backward tolerance:
   *  absent or invalid → 0. (M8) */
  crossfadeSeconds?: number;
}

function isRestorableItem(item: unknown): item is SnapshotItemBase {
  return (
    typeof item === 'object' &&
    !!item &&
    typeof (item as SnapshotItemBase).id === 'string' &&
    ((item as SnapshotItemBase).id as string).length > 0 &&
    !!(item as SnapshotItemBase).source &&
    typeof (item as SnapshotItemBase).source?.kind === 'string' &&
    typeof (item as SnapshotItemBase).source?.url === 'string' &&
    ((item as SnapshotItemBase).source?.url as string).length > 0 &&
    typeof (item as SnapshotItemBase).track?.id === 'number' &&
    !!(item as SnapshotItemBase).track?.audio?.url
  );
}

function clampIndex(index: number, length: number): number {
  return Math.min(Math.max(0, Math.floor(index) || 0), Math.max(0, length - 1));
}

function parseRepeat(v: unknown): RepeatMode {
  return v === 'one' || v === 'all' ? v : 'off';
}

/** Parses a v1 or v2 payload. Returns null when absent/corrupt/empty —
 *  restoration then simply does not happen (never fatal). */
export function decodeSnapshot(raw: string | null | undefined): SessionSnapshot | null {
  if (!raw) return null;
  let candidate: SessionSnapshotV1 | SessionSnapshot | null = null;
  try {
    candidate = JSON.parse(raw) as SessionSnapshotV1 | SessionSnapshot;
  } catch {
    return null;
  }
  if (
    !candidate ||
    (candidate.v !== SNAPSHOT_VERSION && candidate.v !== SNAPSHOT_VERSION_V1) ||
    !Array.isArray(candidate.items)
  ) {
    return null;
  }
  const items = candidate.items.filter(isRestorableItem);
  if (items.length === 0) return null;
  // v1 payloads carry no policy: default to repeat off / shuffle off.
  const legacy = candidate as Partial<SessionSnapshot>;
  // Crossfade: absent/invalid → 0 (off). Only the shipped cycle values are
  // honored (4/8/12); anything else is treated as off.
  const rawCf = legacy.crossfadeSeconds;
  const decoded: SessionSnapshot = {
    v: SNAPSHOT_VERSION,
    items,
    index: clampIndex(candidate.index, items.length),
    positionSeconds: candidate.positionSeconds,
    volume: candidate.volume,
    normalizeMode: candidate.normalizeMode,
    repeat: parseRepeat(legacy.repeat),
    shuffle: legacy.shuffle === true,
    crossfadeSeconds: typeof rawCf === 'number' && [4, 8, 12].includes(rawCf) ? rawCf : 0,
  };
  return decoded;
}

export function encodeSnapshot(snapshot: SessionSnapshot): string {
  return JSON.stringify(snapshot);
}

export { clampIndex };
